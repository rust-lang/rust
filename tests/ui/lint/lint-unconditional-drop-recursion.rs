// Because drop recursion can only be detected after drop elaboration which
// happens for codegen:
//@ build-fail

#![deny(unconditional_recursion)]
#![allow(dead_code)]

pub struct RecursiveDrop;

impl Drop for RecursiveDrop {
    fn drop(&mut self) { //~ ERROR function cannot return without recursing
        let _ = RecursiveDrop;
    }
}

#[derive(Default)]
struct NotRecursiveDrop1;

impl Drop for NotRecursiveDrop1 {
    fn drop(&mut self) {
        // Before drop elaboration, the MIR can look like a recursive drop will
        // occur. But it will not, since forget() prevents drop() from running.
        let taken = std::mem::take(self);
        std::mem::forget(taken);
    }
}

struct NotRecursiveDrop2;

impl Drop for NotRecursiveDrop2 {
    fn drop(&mut self) {
        // Before drop elaboration, the MIR can look like a recursive drop will
        // occur. But it will not, since this will panic.
        std::panic::panic_any(NotRecursiveDrop2);
    }
}

fn main() {}
