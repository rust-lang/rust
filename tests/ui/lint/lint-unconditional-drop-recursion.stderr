error: function cannot return without recursing
  --> $DIR/lint-unconditional-drop-recursion.rs:11:5
   |
LL |     fn drop(&mut self) {
   |     ^^^^^^^^^^^^^^^^^^ cannot return without recursing
LL |         let _ = RecursiveDrop;
   |                              - recursive call site
   |
   = help: a `loop` may express intention better if this is on purpose
note: the lint level is defined here
  --> $DIR/lint-unconditional-drop-recursion.rs:5:9
   |
LL | #![deny(unconditional_recursion)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 1 previous error

